File: mimic/src/progress.f90


subroutine progress_bar(i, total, lenchar, prefix, lenpre)

  implicit none

  integer, intent(in) :: i, total, lenchar, lenpre
  real :: dx
  character :: prefix(lenpre), output(lenchar+lenpre)
  logical :: doprint
  integer :: j

  dx = real(total)/real(lenchar-2)

  doprint = .false.
  if (i .EQ. 1) then
    doprint = .true.
  else if (i .EQ. total) then
    doprint = .true.
  else if (floor((i-1)/dx) .NE. floor((i)/dx)) then
    doprint = .true.
  end if

  do j=1, lenpre
    output(j) = prefix(j)
  end do

  if (doprint .EQV. .true.) then
    output(lenpre+1) = "|"
    do j = 1, lenchar-2
      if (floor(i/dx) .LT. j) then
        output(lenpre+j+1) = "-"
      else
        output(lenpre+j+1) = "#"
      end if
    end do
    output(lenpre+lenchar) = "|"
    print *, output
  end if

end subroutine progress_bar

File: mimic/src/src/progress.py
import numpy as np


def progress_bar(i, total, lenchar, prefix):
    """Display a simple text progress bar to stdout.

    Parameters
    ----------
    i : int
        Current iteration index.
    total : int
        Total number of iterations.
    lenchar : int
        Total width of the printed progress bar string.
    prefix : str
        Prefix string printed before the bar.
    """
    dx = float(total) / float(lenchar - 2)
    doprint = False

    if i == 1 or i == total:
        doprint = True
    else:
        if int((i - 1) / dx) != int(i / dx):
            doprint = True

    if doprint:
        bar_width = max(0, lenchar - 2)
        filled = int(np.floor(i / dx)) if dx != 0 else bar_width
        if filled < 0:
            filled = 0
        if filled > bar_width:
            filled = bar_width
        bar = '#' * filled + '-' * (bar_width - filled)
        print(f"{prefix}|{bar}|")
